core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -116,7 +117,7 @@ public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset)
             throw new IOException("Failed to create a temporary file in "+ dir,e);
         }
 
-        core = Files.newBufferedWriter(tmpPath, charset);
+        core = new FileChannelWriter(tmpPath, charset, false, StandardOpenOption.WRITE);
```

Comment:
Would it helpful to just write to an in-memory buffer? These files are typically smallish, after all, so we do not care about heap usage much, and then we do not care at all who is `flush`ing when. If and when `commit` is called, write the entire buffer to the temp file (if you are lucky this will be a single block), make sure it is synched, then do the atomic rename.

I also think we should add a flag to `XmlFile` allowing callers to opt out of expensive consistency guarantees, with the default being strict mode. `Run.getDataFile` and `SimpleXStreamFlowNodeStorage.getNodeFile` could then use the lax mode, accepting the risk of corrupted data in exchange for better performance. You can even guide callers to make this decision explicitly by (say) deprecating existing `XmlFile` constructors in favor of ones taking a `boolean` flag.

Reply:
> Would it helpful to just write to an in-memory buffer? These files are typically smallish, after all, so we do not care about heap usage much, and then we do not care at all who is flushing when. If and when commit is called, write the entire buffer to the temp file (if you are lucky this will be a single block), make sure it is synched, then do the atomic rename.

sounds a little like an over eager optimisation to me.
When writing bytes to a file (as we do here without the `SYNC` / `DSYNC`  flag) the OS will generally buffer them (almost all OSes have a disk cache buffer) , and they are not written directly to disk.  
The only sync call is the one in commit in the case that `false` is passed (which it is) for AtomicFW you would always want false (as until commit is called it can be thrown away), for any direct uses of the channel writer you may want to know writes have been flushed to disk before returning.